keyboard/src/lib.rs

```diff
@@ -1,11 +1,321 @@
+        0x2e => KC,
+        0x20 => KD,
+        0x12 => KE,
+        0x21 => KF,
+        0x22 => KG,
+        0x23 => KH,
+        0x17 => KI,
+        0x24 => KJ,
+        0x25 => KK,
+        0x26 => KL,
+        0x32 => KM,
+        0x31 => KN,
+        0x18 => KO,
+        0x19 => KP,
+        0x10 => KQ,
+        0x13 => KR,
+        0x1f => KS,
+        0x14 => KT,
+        0x16 => KU,
+        0x2f => KV,
+        0x11 => KW,
+        0x2d => KX,
+        0x15 => KY,
+        0x2c => KZ,
+        0x0b => K0,
+        0x02 => K1,
+        0x03 => K2,
+        0x04 => K3,
+        0x05 => K4,
+        0x06 => K5,
+        0x07 => K6,
+        0x08 => K7,
+        0x09 => K8,
+        0x0a => K9,
+        0x29 => KBacktick,
+        0x0c => KDash,
+        0x0d => KEqual,
+        0x2b => KBackslash,
+        0x39 => KSpace,
+        0x1a => KLeftSquareBracket,
+        0x1b => KRightSquareBracket,
+        0x27 => KSemicolon,
+        0x28 => KApostrophe,
+        0x33 => KComma,
+        0x34 => KPeriod,
+        0x35 => KSlash,
+        0x37 => KKeypadAsterisk, // Keypad
+        0x4a => KKeypadMinus,    // Keypad
+        0x4e => KKeypadPlus,     // Keypad
+        0x2A => KLeftShift,
+        0x36 => KRightShift,
+        _    => return None,
+    };
+
+    Some((key, action))
+}
+
+const MAX_KEYS: usize = 256;
+
+/// PS/2 keyboard driver: manages receiving signals from keyboard port,
+/// updating its internal state according to those signals, and providing
+/// relevant output (if any) based on its state.
+pub struct Keyboard {
+    stack: [Option<Key>; MAX_KEYS],
+    size: usize,
+}
+
+impl Keyboard {
+    pub fn new() -> Keyboard {
+        Keyboard {
+            stack: [None; MAX_KEYS],
+            size: 0,
+        }
+    }
+
+    fn stack_top(&self) -> Option<Key> {
+        if self.size == 0 {
+            None
+        } else {
+            self.stack[self.size - 1]
+        }
+    }
+
+    // Update the driver's internal state in response to a scanned key code
+    pub fn input(&mut self, scancode: u8) -> Option<char> {
+        let key_action = match from_scancode(scancode) {
+            Some(k) => k,
+            None => return None,
+        };
+
+        let (key, action) = key_action;
+
+        match action {
+            Action::Pressed => {
+                self.push_key(key);
+
+                if key.is_printable() {
+                    return self.printable()
+                }
+            },
+            Action::Released => {
+                self.clear_key(key);
+            },
+        }
 
-pub fn from_scancode(code: usize) -> Option<char> {
-    if code < 59 { 
-        Some(CODES[code] as char)
-    } else {
         None
     }
+
+    fn push_key(&mut self, key: Key) {
+        // Don't repeat-register key presses
+        if self.is_key_pressed(key) {
+            return
+        }
+
+        // Guard against overflow
+        if self.size == MAX_KEYS {
+            // TODO: Panic!
```

Comment:
Implemented in 46afc2a. I just realized that if that Just Works then it will be the first kernel panic, literally, I've ever written. :joy: